OsmAnd-java/src/main/java/net/osmand/router/MissingMapsCalculator.java

```diff
@@ -93,24 +95,25 @@ public boolean checkIfThereAreMissingMaps(RoutingContext ctx, LatLon start, List
 		if (end != null) {
 			addPoint(ctx, knownMaps, pointsToCheck, end);
 		}
-		Set<String> usedMaps = new TreeSet<String>();
-		Set<String> mapsToDownload = new TreeSet<String>();
-		Set<String> mapsToUpdate = new TreeSet<String>();
+		Set<String> usedMaps = new LinkedHashSet<>();
+		Set<String> mapsToDownload = new LinkedHashSet<>();
+		Set<String> missingMaps = new LinkedHashSet<>();
+		Set<String> mapsToUpdate = new LinkedHashSet<>();
 		Set<Long> presentTimestamps = null;
 		for (Point p : pointsToCheck) {
 			if (p.hhEditions == null) {
 				if (p.regions.size() > 0) {
-					mapsToDownload.add(p.regions.get(0));
+					CollectionUtils.addIfNotContains(p.regions.get(0), missingMaps, mapsToDownload);
```

Comment:
why is it needed for set? Just add is good enough

Reply:
Fixed in https://github.com/osmandapp/OsmAnd/pull/19695/commits/09d3b74bb496b4138303a2b3abd03185fcfc9b82